solidity/contracts/bank/Bank.sol

```diff
@@ -0,0 +1,299 @@
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/// @title Bitcoin Bank
+/// @notice Bank is a central component tracking Bitcoin balances. Balances can
+///         be transferred between holders and holders can approve their
+///         balances to be spent by others. Balances in the Bank are updated for
+///         depositors who deposit their Bitcoin into the Bridge and only the
+///         Bridge can increase balances.
+/// @dev Bank is a governable contract and the Governance can upgrade the Bridge
+///      address.
+contract Bank is Ownable {
+    address public bridge;
+
+    /// @notice The balance of a given account in the Bank. Zero by default.
+    mapping(address => uint256) public balanceOf;
+
+    /// @notice The remaining amount of balance a spender will be
+    ///         allowed to transfer on behalf of an owner using
+    ///         `transferBalanceFrom`. Zero by default.
+    mapping(address => mapping(address => uint256)) public allowance;
+
+    /// @notice Returns the current nonce for EIP2612 permission for the
+    ///         provided balance owner for a replay protection. Used to
+    ///         construct EIP2612 signature provided to `permit` function.
+    mapping(address => uint256) public nonce;
+
+    uint256 public immutable cachedChainId;
+    bytes32 public immutable cachedDomainSeparator;
+
+    /// @notice Returns EIP2612 Permit message hash. Used to construct EIP2612
+    ///         signature provided to `permit` function.
+    bytes32 public constant PERMIT_TYPEHASH =
+        keccak256(
+            "Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)"
+        );
+
+    event BalanceTransferred(
+        address indexed from,
+        address indexed to,
+        uint256 amount
+    );
+
+    event BalanceApproved(
+        address indexed owner,
+        address indexed spender,
+        uint256 amount
+    );
+
+    event BalanceIncreased(address indexed owner, uint256 amount);
+
+    event BalanceDecreased(address indexed owner, uint256 amount);
+
+    event BridgeUpdated(address newBridge);
+
+    modifier onlyBridge() {
+        require(msg.sender == address(bridge), "Caller is not the bridge");
+        _;
+    }
+
+    constructor() {
+        cachedChainId = block.chainid;
+        cachedDomainSeparator = buildDomainSeparator();
+    }
+
+    /// @notice Allows the Governance to upgrade the Bridge address.
+    /// @dev The function does not implement any governance delay and does not
+    ///      check the status of the Bridge. The Governance implementation needs
+    ///      to ensure all requirements for the upgrade are satisfied before
+    ///      executing this function.
+    function updateBridge(address _bridge) external onlyOwner {
```

Comment:
Regarding your comment: https://github.com/keep-network/tbtc-v2/pull/84/files?diff=unified&w=0#r775931935, don't we want to prepare some "template" of the update process at this point? I mean something like that:
1. We can introduce `IBridge` interface which exposes `disable` and `isDisabled` methods (or any other appropriate names). Implementations could emit special alerts to wallets upon `disable` call to force the sweep. The `isDisabled` will return `true` only if `disable` has been called and deposits have been swept.
2. We divide the update process into two steps. First, we can have `beginBridgeUpdate` which triggers the `disable` method on the old bridge.
3. The second part `finalizeBridgeUpdate` checks whether the old bridge `isDisabled` and makes the actual update. 

That's just a loose idea but I think we should make the update process explicit. Currently, it looks like the bank can update the bridge whenever it wants.

Reply:
I have #39 in the queue and I was planning to tackle it there. We also need to implement upgradeability for TBTC vault by at least transferring TBTC token ownership.

Reply:
Sure, got it!